src/materials/ConcreteASREigenstrain.C

```diff
@@ -0,0 +1,278 @@
+                        "Set true to turn ASR reaction dependent tensile strength");
+  params.addParam<Real>(
+      "beta_f_tensile", 0.5, "Fractional residual tensile strength at full ASR reaction");
+
+  params.addParam<unsigned int>("max_its", 30, "Maximum number of sub-newton iterations");
+  params.addParam<bool>(
+      "output_iteration_info", false, "Set true to output sub-newton iteration information");
+  params.addParam<bool>("output_iteration_info_on_error",
+                        false,
+                        "Set true to output sub-newton iteration information when a step fails");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-5, "Relative convergence tolerance for sub-newtion iteration");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-20, "Absolute convergence tolerance for sub-newtion iteration");
+
+  params.addClassDescription(
+      "Computes the scalar volumetric expansion due to alkali-silica reaction. "
+      "Used in conjunction with another model to apply this as an eigenstrain.");
+
+  return params;
+}
+
+ConcreteASREigenstrain::ConcreteASREigenstrain(const InputParameters & parameters)
+  : ConcreteExpansionEigenstrainBase(parameters, "ASR"),
+    _temperature(coupledValue("temperature")),
+
+    _rh(coupledValue("relative_humidity")),
+    _rh_exponent(getParam<Real>("rh_exponent")),
+
+    _expansion(getParam<MooseEnum>("expansion").getEnum<ExpansionType>()),
+
+    _ASR_final_vstrain(getParam<Real>("ASR_vol_expansion")),
+    _tau_c_T0(getParam<Real>("tau_c_T0")),
+    _tau_L_T0(getParam<Real>("tau_L_T0")),
+    _Uc(getParam<Real>("Uc")),
+    _UL(getParam<Real>("UL")),
+    _ASR_T0(getParam<Real>("ASR_T0")),
+    _alpha(getParam<Real>("alpha")),
+
+    _gamma_tensile(getParam<Real>("gamma_tensile")),
+    _gamma_residual(getParam<Real>("gamma_residual")),
+    _beta(getParam<Real>("beta")),
+
+    _ASR_dependent_tensile_strength(getParam<bool>("ASR_dependent_tensile_strength")),
+    _beta_f_tensile(getParam<Real>("beta_f_tensile")),
+
+    _max_its(parameters.get<unsigned int>("max_its")),
+    _output_iteration_info(getParam<bool>("output_iteration_info")),
+    _output_iteration_info_on_error(getParam<bool>("output_iteration_info_on_error")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+
+    _ASR_extent(declareProperty<Real>("ASR_extent")),
+    _ASR_extent_old(getMaterialPropertyOld<Real>("ASR_extent"))
+{
+}
+
+void
+ConcreteASREigenstrain::initQpStatefulProperties()
+{
+  _ASR_extent[_qp] = 0.0;
+  ConcreteExpansionEigenstrainBase::initQpStatefulProperties();
+}
+
+Real
+ConcreteASREigenstrain::computeQpVolumetricStrain()
+{
+  // Use Newton iteration to determine ASR reaction extent at new time step
+  Real scalar = _ASR_extent_old[_qp];
+  unsigned int it = 0;
+  Real residual = 10;
+  Real norm_residual = 10;
+  Real first_norm_residual = 10;
+
+  std::stringstream iter_output;
```

Comment:
@agiorla I'm definitely open to making this match the procedure used for RIVE. It would change the results, probably by a small amount. I would prefer to keep it the way it is at least for this initial commit just because this exactly matches the solid_mechanics implementation. I've put the parameters related to the iterations in the 'advanced' category, and removed them from the test inputs to avoid widespread use of them.